bkbit/scripts/gfftranslator.py

```diff
@@ -4,133 +4,346 @@
+        column_names = [name.strip('\ufeff').strip() for name in reader.fieldnames]  # Remove invisible characters
+        validate_column_names(column_names, GENOME_ANNOTATION_FILE_COLUMN_NAMES)
+        
+        for row in reader:
+            row = {k.strip('\ufeff').strip(): v.strip('\ufeff').strip() for k, v in row.items()}
+            filter_and_parse_gff3(**row, data_dir=data_dir, output_dir=output_dir, genomeAssemblyObjects=genomeAssemblyObjects, hash_functions=hash_functions)
+            
 
+def filter_and_parse_gff3(assembly_identifier_prefix, assembly_local_unique_identifier, label, taxon_identifier_prefix, taxon_local_unique_identifier, authority, version, gene_identifier_prefix, description, url, data_dir, output_dir, genomeAssemblyObjects, feature_type_filter=['gene', 'pseudogene', 'ncRNA_gene'], hash_functions=['SHA256']):
     """
-    # filter the dataframe to only genes
-    pred = [x in ['gene','ncRNA_gene','pseudogene'] for x in df['type'] ]
+    Filters and parses a GFF3 file, generates various objects, and serializes them.
+
+    Parameters:
+    assembly_identifier_prefix (str): The prefix for the assembly identifier.
+    assembly_local_unique_identifier (str): The local unique identifier for the assembly.
+    label (str): The label for the genome annotation.
+    taxon_identifier_prefix (str): The prefix for the taxon identifier.
+    taxon_local_unique_identifier (str): The local unique identifier for the taxon.
+    authority (str): The authority for the genome annotation.
+    version (str): The version of the genome annotation.
+    gene_identifier_prefix (str): The prefix for the gene identifier.
+    description (str): The description of the genome annotation.
+    url (str): The URL of the GFF3 file.
+    data_dir (str): The directory to store the downloaded data.
+    output_dir (str): The directory to store the serialized objects.
+    genomeAssemblyObjects (dict): A dictionary of genome assembly objects.
+    feature_type_filter (list, optional): The list of feature types to filter for. Defaults to ['gene', 'pseudogene', 'ncRNA_gene'].
+    hash_functions (list, optional): The list of hash functions to generate checksums. Defaults to ['SHA256'].
+
+    """
+    print(f"AUTHORITY: {authority}, LABEL: {label}, TAXON_LOCAL_UNIQUE_ID: {taxon_local_unique_identifier}, VERSION: {version}, GENE_ID_PREFIX: {gene_identifier_prefix}, URL: {url}")
+    gene_name = ''.join(url.split('/')[-1].split('.')[0:-2])
+    print("GENE NAME: ", gene_name)
+
+    if os.path.isfile(data_dir + '/' + gene_name + '.csv'):
+        print(f"Data from url is already downloaded and saved here: {data_dir + '/' + gene_name + '.csv'}")
+        df = pd.read_csv(data_dir + '/' + gene_name + '.csv') 
+    else:
+        print(f"Downloading and saving data from url here: {data_dir + '/' + gene_name + '.csv'}")
+        df = pd.read_csv(url, sep='\t', comment = "#", header=None, names=GFF3_FILE_COLUMN_NAMES)
```

Comment:
Yeah, I think that's good.

We should probably also check that the version starts with `3`. I don't see any `gff-version` line in the GFF2 specification, but I get the sense that we want to be absolutely sure we are not being given a GFF2 file.

Reply:
On thing to say is that we get what we get and that is not possible to ask for a gff3 files if it doesn't exist.
So, @puja-trivedi - are all the test cases we have work with so far gff3? Can you confirm?
If that is the case, we can say we are standardizing to gff3.

When we get a case that is not gff3, then we will handle that as a new task

Reply:
I'm mostly concerned because I do not believe the final column (effectively the description field) in a GFF2 file follows the same schema as GFF3 (and it looks like GFF2.2 has a different number of columns than GFF3)

[GFF3](http://useast.ensembl.org/info/website/upload/gff3.html)
[GFF2](http://gmod.org/wiki/GFF2)
[GFF2.2](http://mblab.wustl.edu/GTF22.html)

Reply:
If everything we have so far is GFF3, I'm fine building against GFF3 exclusively, but I think the code should throw an exception if it is given something else, so that we know when to schedule the new work Lydia is mentioning.

Reply:
Everything so far is GFF3 (both ensembl and ncbi). Based off the GFF3 documentation, "##gff-version 3" is required to be the first line in a gff3 file so I will check for that. If it is not there I will throw an exception.